ngraph/core/reference/src/runtime/reference/tile.cpp

```diff
@@ -14,34 +14,103 @@
 // limitations under the License.
 //*****************************************************************************
 
+#include <algorithm>
 #include <cmath>
-#include <stdio.h>
+#include <cstdio>
+#include <numeric>
 
 #include "ngraph/check.hpp"
 #include "ngraph/runtime/reference/tile.hpp"
 
 using namespace ngraph;
 
-void runtime::reference::tile(
-    const char* arg, char* out, const Shape& in_shape, const Shape& out_shape, size_t elem_size)
+namespace
+{
+    bool is_axis_upper_bound(const Shape& shape,
+                             std::vector<int64_t>& indices,
+                             size_t& axis,
+                             bool& run)
+    {
+        if (axis-- == 0)
+        {
+            run = false;
+            return false;
+        }
+
+        if (++indices[axis] != shape[axis])
+        {
+            axis = indices.size();
+            return false;
+        }
+
+        indices[axis] = 0;
+        return true;
+    }
+
+    /// \brief Returns a vector containing the results of multiplication of the remaining axes for
+    /// each axis
+    /// If dims has shape(2, 2, 3) then the output vector would be (2*3, 3, 1)
+    ///
+    /// \param[in] dims Shape of the output
+    ///
+    /// \return Vector containing calculated values for each axis.
+    std::vector<int64_t> create_pitches(const Shape& dims)
+    {
+        std::vector<int64_t> pitch;
+        pitch.resize(dims.size());
+        std::partial_sum(dims.rbegin(), dims.rend() - 1, pitch.begin(), std::multiplies<int64_t>());
+        std::reverse(pitch.begin(), pitch.end() - 1);
+
+        pitch.back() = 1;
+        return pitch;
+    }
+}
+
+void runtime::reference::tile(const char* arg,
+                              char* out,
+                              const Shape& in_shape,
+                              const Shape& out_shape,
+                              const size_t elem_size,
+                              const std::vector<int64_t>& repeats)
 {
     Shape in_shape_expanded(in_shape);
     in_shape_expanded.insert(in_shape_expanded.begin(), out_shape.size() - in_shape.size(), 1);
-    CoordinateTransform input_transform(in_shape_expanded);
-    CoordinateTransform output_transform(out_shape);
+    size_t block_size = 0;
+    int64_t num_repeats = 0;
+    bool run = true;
+    std::vector<int64_t> indices(in_shape_expanded.size() - 1, 0);
+    size_t axis(indices.size());
+    const int input_rank = in_shape_expanded.size();
+    const int64_t last_dim = in_shape_expanded[input_rank - 1];
+    const std::vector<int64_t> pitches = create_pitches(out_shape);
+    const char* copy = nullptr;
 
-    for (const Coordinate& output_coord : output_transform)
+    while (run)
```

Comment:
I really dislike such infinite cycles.
Looks like we can change the conditional (we change run only `if axis-- == 0`.

Reply:
I can check the axis value with get_axis() method from Counter struct instead of using run variable, but I would have to change the axis type to int64_t so it can take negative values, because with current type (size_t) there is a problem with last decrement.